### PR TITLE
[changelog] Use serializer cache to improve changecapture user performance

### PR DESCRIPTION
**Pull request by ZacAttack**
## [changelog] Use serializer cache to improve changecapture user performance

The deserializer lookup is pretty expensive.  This change should speed things up.

Resolves #XXX

## How was this PR tested?
normal test suite

## Does this PR introduce any user-facing changes?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Make sure to explain your proposed changes and call out the behavior change.

### CODE REVIEW

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java [hunk 1]
```diff
@@ -146,6 +148,8 @@ public VeniceChangelogConsumerImpl(
       this.userEventChunkingAdapter = GenericChunkingAdapter.INSTANCE;
       this.storeDeserializerCache = new AvroStoreDeserializerCache<>(storeRepository, storeName, true);
     }
+
+    avrostoreDeserializerCache = new AvroStoreDeserializerCache(storeRepository, storeName, true);
```
Comment (sixpluszero) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java:152]: looks good to me, just a nit: you could move this line up, and line 149 can reuse the same cache object for generic client?

**Review by sixpluszero — DISMISSED** [review]

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java (hunk 1 above)
Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java:152]: ugh!  Great point.

Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java]: fixed.

Comment (sixpluszero) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java]: Uh I don't think your PR will pass the unit test coverage. As I think last run should already failed on this https://github.com/linkedin/venice/actions/runs/7704830781

Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java]: on it.  This should pass

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java [hunk 2]
```diff
@@ -761,9 +762,7 @@ private PubSubMessage<K, ChangeEvent<V>, VeniceChangeCoordinate> convertChangeEv
   }
 
   private V deserializeValueFromBytes(ByteBuffer byteBuffer, int valueSchemaId) {
-    Schema currentValueSchema = schemaReader.getValueSchema(valueSchemaId);
-    RecordDeserializer<V> valueDeserializer =
-        FastSerializerDeserializerFactory.getFastAvroGenericDeserializer(currentValueSchema, currentValueSchema);
+    RecordDeserializer<V> valueDeserializer = storeDeserializerCache.getDeserializer(valueSchemaId, valueSchemaId);
```
Comment (FelixGV) [inline clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java:765]: Not related to this optimization, but...

Using the same schema for reader and writer is typically frowned upon... is this the right behavior? I am wondering if we should align it with the behavior of the GenericRecord thin client, which always upconverts to the latest schema (i.e. latest as the reader schema). Writer schema must be whatever the payload came with, of course.

**Review by FelixGV — COMMENTED** [review]
Change itself looks good but I left a more general question...

File: clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java (hunk 2 above)
Comment (ZacAttack) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java:765]: That makes sense, I can take that change in another PR (Ads team is a bit hot to try this out to determine if we need to do something more dramatic to get the performance better).  I'll file a BB

Comment (FelixGV) [reply clients/da-vinci-client/src/main/java/com/linkedin/davinci/consumer/VeniceChangelogConsumerImpl.java:765]: LFG! 🚀

**Review by FelixGV — APPROVED** [review]

**Pull request merged**